### ISSUE TITLE
[Bug]: Blind or incorrect Disk Latency reporting 

### ISSUE THREAD

**Issue by stephane-maillan** [netdata/netdata#14089]
### Bug description

**First of all Infinite thanks for your amazing Tool**  

### Issue Overview
Some Disks Latency counters seems to get a buit-in precision collecting issue : 

While doing disk IO on low latency block device we are not able to see : 

- Latency below 1ms 
- Latency between 1ms to 2ms and so on  

### Example

- For latency below 1ms such as 0.8 ms i will see 0 ms in netdata 

- For latency between two "integer" latency such as 1.82ms i will see 1.0 ms in netdata

--> It seem that no "flotting" precision is available

For a 2.76ms Latency we see 2ms in the dashboard and the same in the exporting 

tcpdump 
netdata.127392-r630.disk_await.nvmesh_sdc1_raid10_s5_crc.reads 2 1670285890
netdata.127392-r630.disk_await.nvmesh_sdc1_raid10_s5_crc.reads 2 1670285891

### Consequences

- User are blind / one-eyed / misleaded for all low latency block device such as nvme or low latency SDS that are common today.

- Globally all block device get this issue but it should be less noticeable for the one that are not low latency

- User is not able to rely on netdata disk latency reporting / all other correlation, dahsboarding as so are impacted by using it

### Identified counter

At least the impacted counter is **disk_await**  handled by **st_await in collectors/proc.plugin/proc_diskstats.c**

                    d->rd_await_reads  = rrddim_add(d->st_await, "reads",  NULL,  1, 1, RRD_ALGORITHM_ABSOLUTE);
                    d->rd_await_writes = rrddim_add(d->st_await, "writes", NULL, -1, 1, RRD_ALGORITHM_ABSOLUTE);**

### Issue resolution ?

I looked at the code to calculate the value all seem's correct here 

                rrddim_set_by_pointer(d->st_await, d->rd_await_reads,  (reads  - last_reads)  ? (readms  - last_readms)  / (reads  - last_reads)  : 0);
                rrddim_set_by_pointer(d->st_await, d->rd_await_writes, (writes - last_writes) ? (writems - last_writems) / (writes - last_writes) : 0);

So i guess first that the issue is not in the value computing

I am not fluent with the notion of multiplier and divider 
Anyway it seems that you use these notion in rrddim_add and asked myself if this is not a build in way to report "floating value" by applying a multiplier in the collect then a divider in dashboard ? 

if so this might be a way of resolution ? 

### Other related counter 

**collectors/proc.plugin/proc_diskstats.c**

It seem that other counter might suffer the same issue such as the following (not exhausitive)

**disk_svctm       :  st_svctm**

                   d->rd_svctm_svctm = rrddim_add(d->st_svctm, "svctm", NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);

**disk_ext_await : st_ext_await**

                   d->rd_await_discards = rrddim_add(d->st_ext_await, "discards", NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);
                   d->rd_await_flushes = rrddim_add(d->st_ext_await, "flushes", NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE);

**disk_iotime      : st_iotime**

                d->rd_iotime_reads  = rrddim_add(d->st_iotime, "reads",  NULL,  1, 1, RRD_ALGORITHM_INCREMENTAL);
                d->rd_iotime_writes = rrddim_add(d->st_iotime, "writes", NULL, -1, 1, RRD_ALGORITHM_INCREMENTAL);

**disk_ext_iotime : st_ext_iotime**

               d->rd_iotime_discards = rrddim_add(d->st_ext_iotime, "discards", NULL, 1, 1, RRD_ALGORITHM_INCREMENTAL);
               d->rd_iotime_flushes = rrddim_add(d->st_ext_iotime, "flushes", NULL, 1, 1, RRD_ALGORITHM_INCREMENTAL);

If the issue is located with the precision of the multiplier / divider then Globally all diskstats counter working with latency and not using multiplier / divider might be impacted

### Expected behavior

I am expecting to see correct disk block  read/write latency such as

-  0.9 ms
- 1.1ms
- 10.54 ms 

Correct latency such as what is reporting by other tool like iostat 

### Steps to reproduce

1.Execute IO on a low latency block device such as NVME / SSD / ramdrive (ie with fio)

fio --name=4krandw --ioengine=libaio --blocksize=4k --readwrite=randread --numjobs=1 --iodepth=1 --direct=1 --group_reporting --time_based --runtime=30  --fsync=1  --filename=/dev/$REPLACE_HERE

2.Look at netdata disk_await and compare the value with other tool such as iostat  
You will not be able to see latency below 1ms nor intermediate latency between interger numbered latency

### Installation method

native binary packages (.deb/.rpm)

### System info

```shell
/etc/lsb-release:DISTRIB_ID=Ubuntu
/etc/lsb-release:DISTRIB_RELEASE=20.04
/etc/lsb-release:DISTRIB_CODENAME=focal
/etc/lsb-release:DISTRIB_DESCRIPTION="Ubuntu 20.04.1 LTS"
/etc/os-release:NAME="Ubuntu"
/etc/os-release:VERSION="20.04.1 LTS (Focal Fossa)"
/etc/os-release:ID=ubuntu
/etc/os-release:ID_LIKE=debian
/etc/os-release:PRETTY_NAME="Ubuntu 20.04.1 LTS"
/etc/os-release:VERSION_ID="20.04"
/etc/os-release:VERSION_CODENAME=focal
/etc/os-release:UBUNTU_CODENAME=focal
```

### Netdata build info

```shell
netdata -W buildinfo
Version: netdata v1.37.1
Configure options:  '--build=x86_64-linux-gnu' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--disable-silent-rules' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--libexecdir=${prefix}/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--prefix=/usr' '--sysconfdir=/etc' '--localstatedir=/var' '--libdir=/usr/lib' '--libexecdir=/usr/libexec' '--with-user=netdata' '--with-math' '--with-zlib' '--with-webdir=/var/lib/netdata/www' '--disable-dependency-tracking' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -fdebug-prefix-map=/usr/src/netdata=. -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fdebug-prefix-map=/usr/src/netdata=. -fstack-protector-strong -Wformat -Werror=format-security'
Install type: binpkg-deb
    Binary architecture: x86_64
    Packaging distro:  
Features:
    dbengine:                   YES
    Native HTTPS:               YES
    Netdata Cloud:              YES 
    ACLK:                       YES
    TLS Host Verification:      YES
    Machine Learning:           YES
    Stream Compression:         YES
Libraries:
    protobuf:                YES (system)
    jemalloc:                NO
    JSON-C:                  YES
    libcap:                  NO
    libcrypto:               YES
    libm:                    YES
    tcalloc:                 NO
    zlib:                    YES
Plugins:
    apps:                    YES
    cgroup Network Tracking: YES
    CUPS:                    YES
    EBPF:                    YES
    IPMI:                    YES
    NFACCT:                  YES
    perf:                    YES
    slabinfo:                YES
    Xen:                     NO
    Xen VBD Error Tracking:  NO
Exporters:
    AWS Kinesis:             NO
    GCP PubSub:              NO
    MongoDB:                 NO
    Prometheus Remote Write: YES
Debug/Developer Features:
    Trace Allocations:       NO
```

**Comment by ilyam8** [comment]
Hi, @stephane-maillan. All timings metrics exposed by the Kernel are in milliseconds, see [procfs-diskstats](https://www.kernel.org/doc/Documentation/ABI/testing/procfs-diskstats).

**Comment by ilyam8** [comment]
Fixing this will break exporting with `?sources=as-collected`. I think that is a blocker for now. Will be fixed in Netdata v2.0 (where we will introduce a lot of breaking (likely not backward-compatible) changes).

**Comment by stephane-maillan** [comment]
Hi , thanks for your fast answer

I am not sure the lose of précision is at the time of curTime - prevTime / curOps - prevOps

But more when i follow the code this is 

                rrddim_set_by_pointer(d->st_await, d->rd_await_reads,  (reads  - last_reads)  ? (readms  - last_readms)  / (reads  - last_reads)  : 0);
                rrddim_set_by_pointer(d->st_await, d->rd_await_writes, (writes - last_writes) ? (writems - last_writems) / (writes - last_writes) : 0);

The formula seem just fine i guess ,  and i followed this in /procs/diskstats and all seem fine 

In regards to the code i am not sure if this can be related to integer division in the C code 
Is i recall nice dividing 2 integer give an integer and if we assign the result to a float we will get something like IntegerResult.0
and so lose all the precision

Is it not the purpose of rrddim_add to handle such usecase by applying a divider first so there is a division of floating point and the result is directly a floating  that get again multiplicated to get the final correct value ? (assuming that the divider itself return a float ....)

rd = rrddim_add(st, "id", "name", multiplier, divider, algorithm);

Is so does such change can make the job in the idea :
ie 

                    d->rd_await_reads  = rrddim_add(d->st_await, "reads",  NULL,  1024, 1024, RRD_ALGORITHM_ABSOLUTE);
                    d->rd_await_writes = rrddim_add(d->st_await, "writes", NULL, -1024, 1024, RRD_ALGORITHM_ABSOLUTE);


Or i might mistake  if i look deeper for instance for the disk_avgsz counter it seem that the multiplier and divider are applied only in the dashboarding but that the exported netdata value is not modified at first instance ?!
If so , is it this that you refer to a breaking change as a way to correct this might be for example to expose the exported value as micro / nanosecond or with some kind of multiplier that everybody will have to then divide in their own dashboard (ie graphana)

thanks for your comment

Best Regards

**Comment by ilyam8** [comment]
@stephane-maillan fixed in #18040
